sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -653,170 +664,406 @@ private void refreshAllValuesFromRestrictions(Map<Resource, URI> restrictions) t
+            // Combine all of a type's intersections together.
+            for (final Set<Resource> intersection : intersectionList) {
+                otherTypes.addAll(intersection);
+            }
+            for (final Resource other : otherTypes) {
+                // :A intersectionOf[:B, :C] implies that
+                // :A subclassOf :B
+                // :A subclassOf :C
+                // So add each type that's part of the intersection to the
+                // subClassOf graph.
+                addSubClassOf(type, other);
+                for (final Set<Resource> intersection : intersectionList) {
+                    if (!intersection.contains(other)) {
+                        addIntersection(intersection, other);
+                    }
+                }
+            }
+            for (final Set<Resource> intersection : intersectionList) {
+                addIntersection(intersection, type);
+            }
+        }
+        for (final Entry<Resource, List<Set<Resource>>> entry : intersectionsProp.entrySet()) {
+            final Resource type = entry.getKey();
+            final List<Set<Resource>> intersectionList = entry.getValue();
+
+            final Set<URI> superClasses = getSuperClasses((URI) type);
+            for (final URI superClass : superClasses) {
+                // Add intersections to super classes if applicable.
+                // IF:
+                // :A intersectionOf[:B, :C]
+                // AND
+                // :A subclassOf :D
+                // Then we can infer:
+                // intersectionOf[:B, :C] subclassOf :D
+                for (final Set<Resource> intersection : intersectionList) {
+                    addIntersection(intersection, superClass);
+                }
+            }
+            // Check if other keys have any of the same intersections and infer
+            // the same subclass logic to them that we know from the current
+            // type. Propagating up through all the superclasses.
+            for (final Set<Resource> intersection : intersectionList) {
+                final Set<Resource> otherKeys = Sets.newHashSet(intersectionsProp.keySet());
+                otherKeys.remove(type);
+                for (final Resource otherKey : otherKeys) {
+                    if (intersectionsProp.get(otherKey).contains(intersection)) {
+                        for (final URI superClass : superClasses) {
+                            addSubClassOf(otherKey, superClass);
+                        }
+                    }
+                }
+            }
+        }
+        for (final Entry<Resource, List<Set<Resource>>> entry : intersectionsProp.entrySet()) {
+            final Resource type = entry.getKey();
+            final List<Set<Resource>> intersectionList = entry.getValue();
+            for (final Set<Resource> intersection : intersectionList) {
+                addIntersection(intersection, type);
```

Comment:
Isn't this last loop redundant since we're calling addIntersection(intersection, type) in the first pass (line 728)?